### DISCUSSION TITLE
Portable executable for a Django app

### DISCUSSION THREAD

**Discussion by patoch01220** [pyinstaller/pyinstaller#5911] [Help]
Hello, 
I would like to create and executable of my really simple Django app. 

At first I created a python script that launched the server and opened a webbrowser on the right page.  And then i've tried to use pyInstaller to get the executable for this script so that a person that doesn't have python3 can just click on this executable.
The issue i have is that as I had the command ```python3 manage.py runserver```  it still needed to have python3 installed.

So people suggested me to use pyInstaller on manage.py directly (just as this page says : https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Executable-From-Django).

So i follwed all those steps but it still doesn't work.
When i run the command : ```./dist/TestMACB2/TestMACB2 runserver --noreload``` it says me that there is no module named 'MACB'.

My project name is Project_MACB and the name of my folder that contains the models and so is MACB. 
There is the exact error : 
```
Traceback (most recent call last):
  File "manage.py", line 23, in <module>
    main()
  File "manage.py", line 19, in main
    execute_from_command_line(sys.argv)
  File "django/core/management/__init__.py", line 419, in execute_from_command_line
  File "django/core/management/__init__.py", line 395, in execute
  File "django/__init__.py", line 24, in setup
  File "django/apps/registry.py", line 91, in populate
  File "django/apps/config.py", line 212, in create
  File "importlib/__init__.py", line 127, in import_module
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'MACB'
```

I don't understand how to fix, do you have any idea ?

Thanks for your time :)

**Comment by bwoodsend** [comment]
Is `MACB` importable from wherever you ran PyInstaller? i.e. does `python3 -c "import MACB"` raise a `ModuleNotFoundError`? If it's importable, then just add `--hiddenimport=MACB` to your pyinstaller command. If it isn't then you'll need to pass the `MACB`'s parent folder to the `--paths` option.

**Reply by patoch01220** [reply]
Hey, thanks for your reply! 
I've tried what you said and it seems like i can import it. The issue i have now is : if I run the command : ```./dist/TestMACB2/TestMACB2 runserver``` it tells me ```RuntimeError: Script runserver does not exist.```
And if I run  the command ```./dist/TestMACB2/TestMACB2 runserver --noreload``` It tells me ```ModuleNotFoundError: No module named 'MACB.apps'``` Any ideas ?

**Reply by rokm** [reply]
You need the `--noreload` because PyInstaller-frozen scripts do not support the auto-restart/reload functionality.

> ModuleNotFoundError: No module named 'MACB.apps'

That's probably yet another `hiddenimport` to add... (this applies to most of `ModuleNotFoundError`s). 

If you keep getting many of these, you may want to create a custom hook for MACB: create directory `extra-hooks` and file `extra-hooks/hook-MACB.py` with the following contents:
```python
from PyInstaller.utils.hooks import collect_submodules
hiddenimports = collect_submodules('MACB')
```
and then add `--additional-hooks-dir extra-hooks`  argument to your PyInstaller command-line.

**Reply by patoch01220** [reply]
> You need the `--noreload` because PyInstaller-frozen scripts do not support the auto-restart/reload functionality.
> 
> > ModuleNotFoundError: No module named 'MACB.apps'
> 
> That's probably yet another `hiddenimport` to add... (this applies to most of `ModuleNotFoundError`s).
> 
> If you keep getting many of these, you may want to create a custom hook for MACB: create directory `extra-hooks` and file `extra-hooks/hook-MACB.py` with the following contents:
> 
> ```python
> from PyInstaller.utils.hooks import collect_submodules
> hiddenimports = collect_submodules('MACB')
> ```
> 
> and then add `--additional-hooks-dir extra-hooks` argument to your PyInstaller command-line.

Thank you for your answer! I've tried including both : MACB and MACB.apps and now i have a pretty weird error while trying to run the server : ```INTEL MKL ERROR: dlopen(/Users/patoch/Desktop/Patoch/3rd_Year/2nd_Semester/Travail_Bachelor/Django/Project_MACB/dist/TestMACB2/libmkl_intel_thread.dylib, 9): Library not loaded: @rpath/libiomp5.dylib
  Referenced from: /Users/patoch/Desktop/Patoch/3rd_Year/2nd_Semester/Travail_Bachelor/Django/Project_MACB/dist/TestMACB2/libmkl_intel_thread.dylib
  Reason: image not found.
Intel MKL FATAL ERROR: Cannot load libmkl_intel_thread.dylib.```
Do you have any ideas ?

**Reply by rokm** [reply]
Looks like `libmkl_intel_thread.dylib` requires `libiomp5.dylib` that was not collected (or was collected in the wrong place). What's the source of these shared libraries?

**Reply by patoch01220** [reply]
> Looks like `libmkl_intel_thread.dylib` requires `libiomp5.dylib` that was not collected (or was collected in the wrong place). What's the source of these shared libraries?

I don't know actually, i've never seen them or heard of them. There is an executable ```libmkl_intel_thread.dylib```in the dist folder but I can't say much about those files, i told you i've never heard about them. 

Also I have some warnings when i execute de PyInstaller command. Those warning are about hidden imports (of Django libs mainly) that aren't found. Could it be due to that ? 

There, is few example of the Warnings : 
```WARNING: Hidden import "django.contrib.gis.utils.context_processors" not found!```
```30750 WARNING: Hidden import "pkg_resources.py2_warn" not found!```
```208457 WARNING: Unable to find package for requirement zope.event from package gevent.```
```208457 WARNING: Unable to find package for requirement zope.interface from package gevent.```
```208458 WARNING: Unable to find package for requirement greenlet from package gevent.```
```210164 WARNING: library user32 required via ctypes not found```

**Comment by patoch01220** [comment]
Okay, i have solve my problem by updating numpy. I then add some other small issue that i succeed resolving but now i've got on that I don't understand.

So i succeed creating my executable and it launches the server right. But when i try to access the main page I got this error : 
```
Internal Server Error: /MACB/
Traceback (most recent call last):
  File "django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "MACB/views.py", line 10, in MainView
    return render(request, 'MACB/Main.html', context)
  File "django/shortcuts.py", line 19, in render
    content = loader.render_to_string(template_name, context, request, using=using)
  File "django/template/loader.py", line 61, in render_to_string
    template = get_template(template_name, using=using)
  File "django/template/loader.py", line 19, in get_template
    raise TemplateDoesNotExist(template_name, chain=chain)
django.template.exceptions.TemplateDoesNotExist: MACB/Main.html
[19/Jun/2021 17:18:17] "GET /MACB/ HTTP/1.1" 500 75222
```
Apparently it can't find the template of the main page but if i run the server manually with the command ```python3 manage.py runserver```it works well. Do you have any idea ?

**Reply by rokm** [reply]
> Apparently it can't find the template of the main page but if i run the server manually with the command `python3 manage.py runserver`it works well. Do you have any idea ?

You need to collect the data files from your `MACB` package yourself. If you made a hook for MACB as suggested earlier, you can extend it like this:
```python
from PyInstaller.utils.hooks import collect_submodules, collect_data_files
hiddenimports = collect_submodules('MACB')
datas = collect_data_files('MACB')
```

**Reply by patoch01220** [reply]
Hey, i've tried but i got this error : ```
Traceback (most recent call last):
  File "manage.py", line 9, in <module>
    from PyInstaller.utils.hooks import collect_submodules, collect_data_files
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "PyInstaller/loader/pyimod03_importers.py", line 540, in exec_module
  File "PyInstaller/__init__.py", line 55, in <module>
  File "pkg_resources/__init__.py", line 466, in get_distribution
  File "pkg_resources/__init__.py", line 342, in get_provider
  File "pkg_resources/__init__.py", line 886, in require
  File "pkg_resources/__init__.py", line 772, in resolve
pkg_resources.DistributionNotFound: The 'PyInstaller' distribution was not found and is required by the application```
And apparently this is due to the fact that pyinstaller is not yet supported on python 3.8. Is that right ?

**Reply by rokm** [reply]
>  And apparently this is due to the fact that pyinstaller is not yet supported on python 3.8. Is that right ?

Nope, python 3.8 (and 3.9) are supported and have been for a while now.

But you've put the lines that I gave you into wrong file (`manage.py`, by the look of it). Those go into a hook (`extra-hooks/hook-MACB.py` - look at earlier replies), not into your code.

**Reply by patoch01220** [reply]
Okay i figured it out : I did create the folder and file your said. It was just missing the ```=```in ```--additional-hooks-dir=extra-hooks``` as you said earlier.

**Comment by patoch01220** [comment]
@rokm But now i have a new issue : It seems to work but ii doesn't find my CSS file : 
```
[22/Jun/2021 16:33:05] "GET /MACB/ HTTP/1.1" 200 1820
Not Found: /static/MACB/StyleMain.css
[22/Jun/2021 16:33:05] "GET /static/MACB/StyleMain.css HTTP/1.1" 404 2148
```

Any ideas ? 
(The page does open but just as a white page with my page without any style)

**Reply by rokm** [reply]
You'll need to collect the CSS file yourself, then (are static resources located outside of the `MACB` directory?)...

**Reply by patoch01220** [reply]
Okay, but how do i do that ? And yeah they are in the MACB directory just like you see here : 

<img width="187" alt="Screenshot 2021-06-22 at 17 25 11" src="https://user-images.githubusercontent.com/62348996/122953181-cf2b6600-d37e-11eb-9401-b0d813dba043.png">

Or should I specify it in another way in my HTML script ? I'm currently specifying them like this 
```
<link rel="stylesheet" href="{% static 'MACB/StyleAnalysedScores_3_6.css' %}">
```

**Reply by rokm** [reply]
Are they collected inside `dist/<name>/MACB/static/`?

**Reply by patoch01220** [reply]
Yes just like this 
<img width="363" alt="Screenshot 2021-06-22 at 17 30 47" src="https://user-images.githubusercontent.com/62348996/122954287-9770ee00-d37f-11eb-85bb-d37205a95b5d.png">

**Reply by rokm** [reply]
Then I guess it's not the problem with missing data, but rather with search path.

**Reply by patoch01220** [reply]
> Then I guess it's not the problem with missing data, but rather with search path.

What do you mean by search path ? Do you mean the path that specify the CSS file in the HTML code ?

**Reply by rokm** [reply]
> Do you mean the path that specify the CSS file in the HTML code ?

Yes. Or the prefix that Django uses to construct the path when searching for static resources. Afraid I'm not familiar with django enough to help you there, sorry.

**Reply by patoch01220** [reply]
Okay no problems! Thank you so much for your help until there. It really helped me a lot! Thanks

**Comment by patoch01220** [comment]
Hello again @rokm ahahah

I resolved my issue by including the style directly in the htlm file. 

So now it work well on my computer but i've tried on my sister's computer and I got this error : 
![image](https://user-images.githubusercontent.com/62348996/123508847-29c00d00-d672-11eb-8f1d-0b0f0917652f.png)

Do I really need to install gdal, or could i include it with the executable so she doesn't have to install it ? I mean include it with the command pyInstaller ?

Thanks a lot again !!
